system/datatypes/mailinglist.php

```diff
@@ -0,0 +1,728 @@
+		$table = new HTMLTable();
+		$this->findBySelector('#main')->appendChild($table);
+		$table->addHeaderRow()->addCells(['', __('ml_addresses'), __('ml_status')]);
+		/** @var HyphaDomElement $doc */
+		$doc = $this->getDoc();
+		/** @var HyphaDomElement $addresses */
+		$addresses = $doc->getOrCreate('addresses')->children();
+		foreach ($addresses as $child) {
+			$table->addRow()->addCells(['', $child->getAttribute('email'), $child->getAttribute('status')]);
+		}
+		$table->addRow()->addCells([__('total'), count($addresses), '']);
+	}
+
+	private function createMailingAction() {
+		return $this->editMailingAction();
+	}
+
+	private function editMailingAction($mailingId = null) {
+		// throw error if edit is requested without client logged in
+		if (!isUser()) return notify('error', __('login-to-edit'));
+
+		// get mailing if id was given
+		/** @var HyphaDomElement $mailing */
+		if ($mailingId) {
+			// check if given id was correct
+			$mailing = $this->xml->document()->getElementById($mailingId);
+			if (!$mailing instanceof HyphaDomElement) {
+				return notify('error', __('invalid-id-given'));
+			}
+			// check if status is correct
+			if ($mailing->getAttribute('status') == self::MAILING_STATUS_SENT) {
+				return notify('error', __('unable-to-edit'));
+			}
+		}
+
+		// check if form is posted and get form data
+		$formSaved = $this->isPosted(self::FORM_CMD_MAILING_SAVE);
+		$formSend = $this->isPosted(self::FORM_CMD_MAILING_SEND);
+		$formPosted = $formSaved || $formSend;
+		$formData = [];
+		if ($formPosted) {
+			$formData = $_POST;
+		} elseif ($mailingId) {
+			$formData = [
+				self::FIELD_NAME_ID => $mailingId,
+				self::FIELD_NAME_SENDER => $mailing->getAttribute('sender'),
+				self::FIELD_NAME_SUBJECT => $mailing->getAttribute('subject'),
+				self::FIELD_NAME_MESSAGE => $mailing->getHtml(),
+			];
+		}
+
+		// create form
+		$form = $this->createMailingForm($formData);
+
+		// process form if it was posted
+		if ($formPosted) {
+			$form->validateRequiredField(self::FIELD_NAME_SENDER);
+			$form->validateEmailField(self::FIELD_NAME_SENDER);
+			$form->validateRequiredField(self::FIELD_NAME_SUBJECT);
+			$form->validateRequiredField(self::FIELD_NAME_MESSAGE);
+			if (empty($form->errors)) {
+				// create or update mailing with given data
+				$mailing = $this->createOrUpdateMailing($form->data);
+
+				// send form if send button was used
+				if ($formSend) {
+					// iterate over the confirmed addresses and send the mailing
+					/** @var HyphaDomElement[] $addresses */
+					$addresses = $this->getDoc()->findXPath('//address[@status="' . self::ADDRESS_STATUS_CONFIRMED . '"]');
+					$receivers = [];
+					foreach ($addresses as $address) {
+						$receivers[] = $address->getAttribute('email');
+					}
+					$this->send($mailing->getAttribute('subject'), $mailing->getHtml(), $receivers, $mailing->getAttribute('sender'));
+
+					// mark mailing as sent
+					$this->xml->lockAndReload();
```

Comment:
implemented the sending/sent part.

we would need to think of a 'simple' solution in order to prevent a time out. simple chunking it will not de the trick imo. Something to discuss. I'll add a todo.